File: apps/frontend/src/app/foundation/notifications/notification.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { ApiClient, ApiPage } from '@foundation/networking';
import {
  NotificationItem,
  RegisterDeviceTokenRequest,
  UnreadNotificationCount,
} from './notification.models';

@Injectable({ providedIn: 'root' })
export class NotificationService {
  private readonly api = inject(ApiClient);

  list(page = 0, size = 20): Observable<ApiPage<NotificationItem>> {
    return this.api.get<ApiPage<NotificationItem>>('/notifications', {
      params: {
        page,
        size,
      },
    });
  }

  unreadCount(): Observable<UnreadNotificationCount> {
    return this.api.get<UnreadNotificationCount>('/notifications/unread-count');
  }

  markRead(notificationId: string): Observable<void> {
    return this.api.post<void>(`/notifications/${encodeURIComponent(notificationId)}/read`, {});
  }

  registerDeviceToken(payload: RegisterDeviceTokenRequest): Observable<void> {
    return this.api.post<void, RegisterDeviceTokenRequest>('/notifications/device-tokens', payload);
  }
}

File: apps/ios/SpotLink/Sources/SpotLink/Features/Notifications/NotificationService.swift
import Foundation

public protocol DeviceTokenLifecycleServicing: Sendable {
    func registerDeviceToken(_ token: String) async throws
    func unregisterDeviceToken(_ token: String) async throws
}

// MARK: - Notification Service

public final class NotificationService: DeviceTokenLifecycleServicing {
    private let apiClient: APIClientProtocol

    public init(apiClient: APIClientProtocol) {
        self.apiClient = apiClient
    }

    public func listNotifications(page: Int = 0, size: Int = 20) async throws -> APIPage<SpotLinkNotification> {
        try await apiClient.get("/notifications", query: ["page": String(page), "size": String(size)])
    }

    public func unreadCount() async throws -> Int {
        let response: NotificationUnreadCount = try await apiClient.get("/notifications/unread-count", query: nil)
        return response.count
    }

    public func markRead(_ notificationId: String) async throws {
        let _: EmptyResponse = try await apiClient.post("/notifications/\(notificationId)/read", body: EmptyPayload())
    }

    /// Registruje APNs device token na backendu.
    public func registerDeviceToken(_ token: String) async throws {
        let _: EmptyResponse = try await apiClient.post(
            "/notifications/device-tokens",
            body: RegisterDeviceTokenRequest(deviceToken: token))
    }

    /// Odjavljuje APNs device token sa backenda za trenutnog korisnika.
    public func unregisterDeviceToken(_ token: String) async throws {
        let _: EmptyResponse = try await apiClient.post(
            "/notifications/device-tokens/unregister",
            body: UnregisterDeviceTokenRequest(deviceToken: token))
    }
}

private struct EmptyPayload: Encodable {}
